### ISSUE TITLE
@cds/react: controlled form inputs

### ISSUE THREAD

**Issue by coryrylan** [vmware/clarity#5625]
## Describe the bug

Right now in @cds/react certain form inputs do not work well with controlled input types in React. In this example the second checkbox event is ignored and not updated.

https://stackblitz.com/edit/react-cds-checkbox-onchange-bgcacb?file=index.tsx 

This seems specific to how React handles Controlled vs Uncontrolled input types as the plain html/js works without issue
https://stackblitz.com/edit/typescript-2ysgfw?file=index.html

https://reactjs.org/docs/forms.html#controlled-components 

## How to reproduce

_Please provide a link to a reproduction scenario using one of the [Clarity Stackblitz templates](https://stackblitz.com/@clr-team). Reports without a clear reproduction may not be prioritized until one is provided._

1.  Go to  https://stackblitz.com/edit/react-cds-checkbox-onchange-bgcacb?file=index.tsx 
2.  Click on the first checkbox
3.  Click on the second checkbox
4.  See missing log from second checkbox

## Expected behavior

Controlled inputs should work with the custom element.

## Versions

**Clarity project:**

- [x] Clarity Core
- [ ] Clarity Angular/UI

**Clarity version:**

- [ ] v3.x
- [ ] v4.x
- [x] v5.x

**Framework:**

- [ ] Angular
- [x] React
- [ ] Vue
- [ ] Other:

**Framework version:**
_ie: Angular 11_

**Device:**

- Type: [e.g. MacBook]

## Additional notes

_Add any other notes about the problem here._

**Comment by ashleyryan** [comment]
It looks like this is fixed, do you know when your next release will be?

**Comment by coryrylan** [comment]
@ashleyryan likely sometime this week

**Comment by ashleyryan** [comment]
Hi @coryrylan 

I just updated to the newest version of Clarity Core, and the inputs still have different behavior than vanilla inputs. If I set a CdsCheckbox's value programmatically, the onChange event will fire, even though the user didn't interact with it. A regular input only fires `onChange` when the input is clicked. 

Normally this isn't a huge deal, because if you call `setState` again but the value doesn't change, nothing re-renders, nbd. However, we have a couple of gnarly multiselect/combo box components that assume `onChange` means that the value has changed manually and toggles the results between true/false without checking for the value of `e.target.checked`. This is resulting in an infinite loop that's crashing the browser. For a number of reasons, this behavior isn't so easy to change and we would expect the behavior to mimic that of a normal input element in react.

Updated stackblitz: https://stackblitz.com/edit/react-cds-checkbox-onchange-bgcacb-vowfux?file=index.tsx

If you look in the console, you'll see when the events fire

**Comment by coryrylan** [comment]
Ok, we will need to revisit this. The primary issue we are running into is React overrides the setters on the native input when using controlled inputs. This breaks the native setter call which was causing the original issue. When the user interacts with the input React ignores the native DOM event and skips a render cycle since it uses its own setter and internal event. Here is a demo of the issue

https://stackblitz.com/edit/react-ts-nwvy1e 

https://hustle.bizongo.in/simulate-react-on-change-on-controlled-components-baa336920e04 
https://stackoverflow.com/questions/23892547/what-is-the-best-way-to-trigger-onchange-event-in-react-js#46012210 
https://github.com/facebook/react/issues/11488

**Comment by ashleyryan** [comment]
Also, possibly tangentially related, but if I wrap a `CdsCheckbox` in a label to make a bigger click area and then click on the checkbox itself, the input immediately checks and then unchecks and `onChange` is fired twice. This only happens when clicking directly on the input, not the label.

```
<label
        style={{
          width: "100%",
          padding: 8,
          backgroundColor: "lightblue",
          display: "block"
        }}
      >
        <CdsCheckbox>
          <label>CB 1_2</label>
          <input
            type="checkbox"
            name="twolabels"
            onChange={e => {
              console.log(e.target.checked);
            }}
          />
        </CdsCheckbox>
      </label>
      ```
      
 If I were to try and create my own react component that works and looks exactly like the Clarity one, would I be able to use any of the existing CSS? I suspect not because of all the `::host` selectors.

**Comment by ashleyryan** [comment]
Another example, this time for the radio group. While you can avoid unnecessary re-renders in checkboxes by making sure to use the `e.target.checked` property, radios are going to result in multiple re-renders since `onChange` ends up being called n+1 times, with different values

https://user-images.githubusercontent.com/9469374/110135319-5e931600-7d9c-11eb-9a35-f670cbc4ead7.mov

**Comment by ashleyryan** [comment]
I accidentally ended up in a twitter thread about React Components + Web Components and one of the maintainers from LitElement asked me about the issues we were facing, so I described the issue and he agreed that controlled components are tricky and suggested a "controlled" component: https://twitter.com/justinfagnani/status/1369071959175090176

I was playing around in Core and tried something like this in the update event of `control-inline.element.ts`, but then it just ended up immediately undoing my selection since the action seemed to be firing twice:

```
    if (this.controlled && props.get('checked') !== undefined && props.get('checked') !== this.checked) {
      this.checked = props.get('checked');
    }
```
    
 I suspect some of the event listeners created in `firstUpdated` need to be tweaked as well, but explicitly telling the web component that it's controlled might be a good approach here. 
 
 Of course, then there's the problem that it's way too easy to leave `controlled` off of `CdsCheckbox` when creating your component. And with react, using the `checked` property instead of `defaultChecked` is already how you denote controlled vs uncontrolled, so this feels like an extra step that shouldn't be necessary.

**Comment by coryrylan** [comment]
@ashleyryan thanks for the update and link to the twitter thread! I'm working on this throughout this week so if I make progress I'll be sure to post back here.

**Comment by coryrylan** [comment]
@ashleyryan I have a potential fix, I am not yet 100% sure this is covering all the use cases. As far as I can test it seems to get the components in sync. https://github.com/vmware/clarity/pull/5718

**Comment by ashleyryan** [comment]
It fixes the stackblitz issue anyway, I copied the contents into App.tsx in the clarity react folder and it works as expected. Also tested out making sure indeterminate still sets like it should.

**Comment by coryrylan** [comment]
Merged #5718 but going to keep this open until its published so we can follow up

**Comment by coryrylan** [comment]
Closing for now, will reopen if needed
